### PR TITLE
libc: Move pthread_key_destructor to task_info_s

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
See #3626 and #3858

This patch move the pthread key destructor from task_group_s to task_info_s (in main task stack) to let destructors can access in userspace.

## Testing
Passed ostest on esp32c3/maix-bit.

### CODE REVIEW

**Comment by gustavonihei** [comment]
@no1wudi Please, fill the required fields in the PR description.

File: include/nuttx/tls.h [hunk 1]
```diff
@@ -63,6 +63,15 @@
 
 struct task_info_s
 {
+  sem_t tg_sem;
+
+#if CONFIG_TLS_NELEM > 0
+  tls_ndxset_t tg_tlsset;                   /* Set of TLS indexes allocated */
+
+  tls_dtor_t  tg_tlsdestr[CONFIG_TLS_NELEM];  /* List of TLS destructors    */
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply include/nuttx/tls.h]: This mistake is made by cherry-pick, fixed

File: libs/libc/tls/Make.defs [hunk 2]
```diff
@@ -21,6 +21,7 @@
 CSRCS += task_getinfo.c
 
 ifneq ($(CONFIG_TLS_NELEM),0)
+CSRCS += tls_alloc.c tls_free.c tls_getdtor.c tls_getset.c tls_setdtor.c
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply libs/libc/tls/Make.defs]: Removed

File: include/nuttx/tls.h [hunk 3]
```diff
@@ -63,6 +63,13 @@
 
 struct task_info_s
 {
+  sem_t tg_sem;
```
Comment (xiaoxiang781216) [inline include/nuttx/tls.h:66]: replace tg_ to ta_

File: include/nuttx/tls.h [hunk 4]
```diff
@@ -247,7 +255,7 @@ FAR struct tls_info_s *tls_get_info(void);
  ****************************************************************************/
 
 #if CONFIG_TLS_NELEM > 0
-int tls_set_dtor(int tlsindex, tls_dtor_t destr);
+int tls_set_dtor(int tlsindex, tls_dtor_t dtor);
```
Comment (xiaoxiang781216) [inline include/nuttx/tls.h:258]: remove and tls_get_dtor

File: libs/libc/tls/tls_getvalue.c [hunk 5]
```diff
@@ -56,7 +56,7 @@
 
 uintptr_t tls_get_value(int tlsindex)
 {
-  FAR struct tls_info_s *info;
+  FAR struct tls_info_s *info = up_tls_info();
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_getvalue.c:59]: don't need change

File: libs/libc/pthread/pthread_keycreate.c [hunk 6]
```diff
@@ -76,20 +76,16 @@ int pthread_key_create(FAR pthread_key_t *key,
...
   if (tlsindex >= 0)
     {
-      /* Yes.. Return the key value and success */
-
-      *key = (pthread_key_t)tlsindex;
-      tls_set_dtor(tlsindex, destructor);
       return OK;
     }
 
-  return -tlsindex;
```
Comment (xiaoxiang781216) [inline libs/libc/pthread/pthread_keycreate.c:92]: revert the change

File: libs/libc/tls/tls_setvalue.c [hunk 7]
```diff
@@ -57,15 +57,14 @@
 
 int tls_set_value(int tlsindex, uintptr_t tlsvalue)
 {
-  FAR struct tls_info_s *info;
+  FAR struct tls_info_s *info = up_tls_info();
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_setvalue.c:60]: don't need change

File: sched/group/group_create.c [hunk 8]
```diff
@@ -180,6 +180,10 @@ int group_allocate(FAR struct task_tcb_s *tcb, uint8_t ttype)
       goto errout_with_stream;
     }
 
+  /* Initial user space semaphore */
+
+  nxsem_init(&group->tg_info->tg_sem, 0, 1);
```
Comment (xiaoxiang781216) [inline sched/group/group_create.c:185]: need call nxsem_destroy when release

File: libs/libc/tls/tls_alloc.c [hunk 9]
```diff
@@ -52,49 +50,54 @@
...
-  spin_unlock_irqrestore(NULL, flags);
+  _SEM_POST(&info->tg_sem);
 
   /* Check if found a valid TLS data index. */
 
   if (candidate < CONFIG_TLS_NELEM)
     {
       /* Yes.. Return the TLS index and success */
 
+      info->tg_tlsdtor[candidate] = dtor;
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_alloc.c:95]: move after line 82

File: libs/libc/tls/tls_alloc.c [hunk 10]
```diff
@@ -52,49 +50,54 @@
...
 
-  /* Search for an unused index.  This is done in a critical section here to
-   * avoid concurrent modification of the group TLS index set.
-   */
+  ret = _SEM_WAIT(&info->tg_sem);
+
+  if (ERROR == ret)
+    {
+      ret = _SEM_ERRVAL(ret);
+      goto errout_with_errno;
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_alloc.c:70]: direct return

File: libs/libc/tls/tls_alloc.c [hunk 11]
```diff
@@ -52,49 +50,54 @@
...
-  FAR struct task_group_s *group = rtcb->group;
-  irqstate_t flags;
+  FAR struct task_info_s *info = task_get_info();
   int candidate;
   int ret = -EAGAIN;
 
-  DEBUGASSERT(group != NULL);
+  DEBUGASSERT(info);
 
-  /* Search for an unused index.  This is done in a critical section here to
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_alloc.c:68]: why remove the comment?

File: libs/libc/tls/tls_alloc.c [hunk 12]
```diff
@@ -52,49 +50,54 @@
...
-   * avoid concurrent modification of the group TLS index set.
-   */
+  ret = _SEM_WAIT(&info->tg_sem);
+
+  if (ERROR == ret)
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_alloc.c:67]: ret <0

File: libs/libc/tls/tls_alloc.c [hunk 13]
```diff
@@ -52,49 +50,54 @@
...
+    {
+      ret = _SEM_ERRVAL(ret);
+      goto errout_with_errno;
+    }
 
-  flags = spin_lock_irqsave(NULL);
```
Comment (xiaoxiang781216) [inline libs/libc/tls/tls_alloc.c:72]: move ret = -EAGAIN from line 61 to here

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.
